core/src/test/java/org/elasticsearch/cluster/metadata/TemplateUpgradeServiceTests.java

```diff
@@ -0,0 +1,330 @@
+            }
+        }
+
+        if (shouldRemove) {
+            assertThat(changes.v2(), hasSize(1));
+            assertThat(changes.v2().contains("removed_test_template"), equalTo(true));
+        } else {
+            assertThat(changes.v2(), empty());
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testUpdateTemplates() {
+        int additionsCount = randomIntBetween(0, 5);
+        int deletionsCount = randomIntBetween(0, 3);
+
+        List<ActionListener<PutIndexTemplateResponse>> putTemplateListeners = new ArrayList<>();
+        List<ActionListener<DeleteIndexTemplateResponse>> deleteTemplateListeners = new ArrayList<>();
+
+        Client mockClient = mock(Client.class);
+        AdminClient mockAdminClient = mock(AdminClient.class);
+        IndicesAdminClient mockIndicesAdminClient = mock(IndicesAdminClient.class);
+        when(mockClient.admin()).thenReturn(mockAdminClient);
+        when(mockAdminClient.indices()).thenReturn(mockIndicesAdminClient);
+
+        doAnswer(invocation -> {
+            Object[] args = invocation.getArguments();
+            assert args.length == 2;
+            PutIndexTemplateRequest request = (PutIndexTemplateRequest) args[0];
+            assertThat(request.name(), equalTo("add_template_" + request.order()));
+            putTemplateListeners.add((ActionListener) args[1]);
+            return null;
+        }).when(mockIndicesAdminClient).putTemplate(any(PutIndexTemplateRequest.class), any(ActionListener.class));
+
+        doAnswer(invocation -> {
+            Object[] args = invocation.getArguments();
+            assert args.length == 2;
+            DeleteIndexTemplateRequest request = (DeleteIndexTemplateRequest) args[0];
+            assertThat(request.name(), startsWith("remove_template_"));
+            deleteTemplateListeners.add((ActionListener) args[1]);
+            return null;
+        }).when(mockIndicesAdminClient).deleteTemplate(any(DeleteIndexTemplateRequest.class), any(ActionListener.class));
+
+
+        Set<String> deletions = new HashSet<>(deletionsCount);
+        for (int i = 0; i < deletionsCount; i++) {
+            deletions.add("remove_template_" + i);
+        }
+        Map<String, BytesReference> additions = new HashMap<>(additionsCount);
+        for (int i = 0; i < additionsCount; i++) {
+            additions.put("add_template_" + i, new BytesArray("{\"index_patterns\" : \"*\", \"order\" : " + i + "}"));
+        }
+
+        TemplateUpgradeService service = new TemplateUpgradeService(Settings.EMPTY, mockClient, clusterService, null,
+            Collections.emptyList());
+
+        service.updateTemplates(additions, deletions);
+
+        for (int i = 0; i < additionsCount; i++) {
+            if (randomBoolean()) {
+                putTemplateListeners.get(i).onFailure(new RuntimeException());
+            } else {
+                putTemplateListeners.get(i).onResponse(new PutIndexTemplateResponse(randomBoolean()) {
+
```

Comment:
its not clear to me what the purpose of this response listener is?  we are not asserting anything when invoking the action listener, so I'm not clear what its role is in the test?